plugin/trino-opa/src/main/java/io/trino/plugin/opa/OpaAccessControl.java

```diff
@@ -0,0 +1,978 @@
+    public void checkCanExecuteQuery(Identity identity)
+    {
+        opaHighLevelClient.queryAndEnforce(OpaQueryContext.fromIdentity(identity), "ExecuteQuery", AccessDeniedException::denyExecuteQuery);
+    }
+
+    @Override
+    public void checkCanViewQueryOwnedBy(Identity identity, Identity queryOwner)
+    {
+        opaHighLevelClient.queryAndEnforce(OpaQueryContext.fromIdentity(identity), "ViewQueryOwnedBy", AccessDeniedException::denyViewQuery, queryOwner);
+    }
+
+    @Override
+    public Collection<Identity> filterViewQueryOwnedBy(Identity identity, Collection<Identity> queryOwners)
+    {
+        return opaHighLevelClient.parallelFilterFromOpa(
+                queryOwners,
+                queryOwner -> buildQueryInputForSimpleResource(
+                        OpaQueryContext.fromIdentity(identity),
+                        "FilterViewQueryOwnedBy",
+                        OpaQueryInputResource.builder().user(new TrinoUser(queryOwner)).build()));
+    }
+
+    @Override
+    public void checkCanKillQueryOwnedBy(Identity identity, Identity queryOwner)
+    {
+        opaHighLevelClient.queryAndEnforce(OpaQueryContext.fromIdentity(identity), "KillQueryOwnedBy", AccessDeniedException::denyKillQuery, queryOwner);
+    }
+
+    @Override
+    public void checkCanReadSystemInformation(Identity identity)
+    {
+        opaHighLevelClient.queryAndEnforce(OpaQueryContext.fromIdentity(identity), "ReadSystemInformation", AccessDeniedException::denyReadSystemInformationAccess);
+    }
+
+    @Override
+    public void checkCanWriteSystemInformation(Identity identity)
+    {
+        opaHighLevelClient.queryAndEnforce(OpaQueryContext.fromIdentity(identity), "WriteSystemInformation", AccessDeniedException::denyWriteSystemInformationAccess);
+    }
+
+    @Override
+    public void checkCanSetSystemSessionProperty(Identity identity, String propertyName)
+    {
+        opaHighLevelClient.queryAndEnforce(
+                OpaQueryContext.fromIdentity(identity),
+                "SetSystemSessionProperty",
+                AccessDeniedException::denySetSystemSessionProperty,
+                OpaQueryInputResource.Builder::systemSessionProperty,
+                propertyName);
+    }
+
+    @Override
+    public boolean canAccessCatalog(SystemSecurityContext context, String catalogName)
+    {
+        return opaHighLevelClient.queryOpaWithSimpleResource(
+                OpaQueryContext.fromSystemSecurityContext(context),
+                "AccessCatalog",
+                OpaQueryInputResource.builder().catalog(catalogName).build());
+    }
+
+    @Override
+    public void checkCanCreateCatalog(SystemSecurityContext context, String catalog)
+    {
+        opaHighLevelClient.queryAndEnforce(
+                OpaQueryContext.fromSystemSecurityContext(context),
+                "CreateCatalog",
+                AccessDeniedException::denyCreateCatalog,
+                OpaQueryInputResource.Builder::catalog,
+                catalog);
```

Comment:
I think between these two I would rather the first, there isn't much benefit to the extra complexity of passing a callable to perform operations on the resource builder.

My original intention with passing in the "raw" object and a method reference to a method in the OpaQueryInputResource.Buidler was so we could abstract the deny logic and the query building logic in the same way

Say
```java
        opaHighLevelClient.queryAndEnforce(
                OpaQueryContext.fromSystemSecurityContext(context),
                "CreateCatalog",
                catalog,
                AccessDeniedException::denyCreateCatalog,
                OpaQueryInputResource.Builder::catalog);
```

In this instance, `denyCreateCatalog` can get the `catalog` argument directly, and so can `OpaQueryInputResource.Builder::catalog`.

My idea didn't quite work out as you pointed out, since this isn't always possible and there are so many possible combinations that it exploded the number of overloads we needed to implement

I will change to the first style, and will also create small helpers where needed directly in `OpaAccessControl` to try to reduce duplication